javalib/src/main/scala/java/util/internal/GenericArrayOps.scala

```diff
@@ -0,0 +1,146 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.util.internal
+
+import java.lang.{reflect => jlr}
+import java.util.Comparator
+
+/** Typeclasses for generic operations on `Array`s. */
+object GenericArrayOps {
+
+  /** A typeclass for operations to manipulate existing arrays. */
+  sealed trait ArrayOps[A] {
+    def length(a: Array[A]): Int
+    def get(a: Array[A], i: Int): A
+    def set(a: Array[A], i: Int, v: A): Unit
+  }
+
+  /** A typeclass for the ability to create arrays of a given type. */
+  sealed trait ArrayCreateOps[A] {
+    def create(length: Int): Array[A]
+  }
+
+  // ArrayOps and ArrayCreateOps instances for reference types
+
+  private object ReusableAnyRefArrayOps extends ArrayOps[AnyRef] {
+    @inline def length(a: Array[AnyRef]): Int = a.length
+    @inline def get(a: Array[AnyRef], i: Int): AnyRef = a(i)
+    @inline def set(a: Array[AnyRef], i: Int, v: AnyRef): Unit = a(i) = v
+  }
+
+  @inline
+  implicit def specificAnyRefArrayOps[A <: AnyRef]: ArrayOps[A] =
+    ReusableAnyRefArrayOps.asInstanceOf[ArrayOps[A]]
+
+  @inline
+  final class ClassArrayOps[A <: AnyRef](clazz: Class[_ <: Array[A]])
+      extends ArrayCreateOps[A] {
+    @inline def create(length: Int): Array[A] =
+      createArrayOfClass(clazz, length)
+  }
+
+  @inline
+  final class TemplateArrayOps[A <: AnyRef](template: Array[A])
+      extends ArrayCreateOps[A] {
+    @inline def create(length: Int): Array[A] =
+      createArrayOfClass(template.getClass(), length)
+  }
+
+  @inline
+  def createArrayOfClass[A <: AnyRef](clazz: Class[_ <: Array[A]], length: Int): Array[A] =
+    jlr.Array.newInstance(clazz.getComponentType(), length).asInstanceOf[Array[A]]
+
+  implicit object AnyRefArrayCreateOps extends ArrayCreateOps[AnyRef] {
+    @inline def create(length: Int): Array[AnyRef] = new Array[AnyRef](length)
+  }
+
+  /* ArrayOps and ArrayCreateOps instances for primitive types.
+   *
+   * With the exception of the one for Boolean, they also implement
+   * `java.util.Comparator` for the same element type. In a perfect design, we
+   * would define separate objects for that, but it would result in more
+   * generated code for no good reason.
+   */
+
+  implicit object BooleanArrayOps
+      extends ArrayOps[Boolean] with ArrayCreateOps[Boolean] {
+    @inline def length(a: Array[Boolean]): Int = a.length
+    @inline def get(a: Array[Boolean], i: Int): Boolean = a(i)
+    @inline def set(a: Array[Boolean], i: Int, v: Boolean): Unit = a(i) = v
+    @inline def create(length: Int): Array[Boolean] = new Array[Boolean](length)
+  }
+
+  implicit object CharArrayOps
+      extends ArrayOps[Char] with ArrayCreateOps[Char] with Comparator[Char] {
```

Comment:
Do we need to be concerned about providing an implicit `Comparator` publicly? If yes, maybe we can make the objects private and provide `@inline` defs to access them as the relevan types?

Reply:
I don't think this is a problem. In the javalib, if we need a comparator for a primitive type, I don't see how it would ever need to be something *different* than the behavior here.

Reply:
Fair enough :)